### PR TITLE
Users: handling the retrieve_password() answer correctly (as potential WP_Error)

### PR DESCRIPTION
**Pull request by LeTraceurSnork**
Handling the `retrieve_password()` in `wp-admin/users.php` answer correctly (as potential WP_Error) to prevent incorrect `Password reset link sent` for users that are not allowed to reset their passwords

Trac ticket: https://core.trac.wordpress.org/ticket/58407

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by petitphp** [comment]
Change looks good to me 👍.

As proposed in the ticket, we could also update the other place where this check is done :
* in [wp-admin/includes/ajax-actions.php](https://core.trac.wordpress.org/browser/tags/6.2.2/src/wp-admin/includes/ajax-actions.php?marks=5554#L5550) : change to use a check with `is_wp_error`

**Comment by LeTraceurSnork** [comment]
@petitphp do you think I should do that in this PR or another PR related to same trac-ticket or should I open the whole new trac-ticket and open a PR for that one?

**Comment by petitphp** [comment]
@LeTraceurSnork you can do that in the current PR, no need for a new PR or trac ticket 🙂

File: .gitignore [hunk 1]
```diff
@@ -40,6 +40,7 @@ wp-tests-config.php
 /packagehash.txt
 /artifacts
 /setup.log
+/.idea
```
Comment (audrasjb) [inline .gitignore]: I guess this is an unwanted change :)

**Review by audrasjb — CHANGES_REQUESTED** [review]
@LeTraceurSnork would you mind adding a commit to the PR to remove the change done in `.gitignore` file? Thanks!

File: .gitignore (hunk 1 above)
Comment (LeTraceurSnork) [reply .gitignore]: Reverted it. But what do we gotta do with the problem then:
IDE PhpStorm (by IntellijIdea) creates .idea/ folder at the repository root, so either I'm ignoring that or selecting files to commit one-by-one which is kinda unhandy

**Review by petitphp — APPROVED** [review]

File: src/wp-admin/includes/ajax-actions.php [hunk 2]
```diff
@@ -5566,7 +5566,7 @@ function wp_ajax_send_password_reset() {
 	$user    = get_userdata( $user_id );
 	$results = retrieve_password( $user->user_login );
 
-	if ( true === $results ) {
+	if ( ! is_wp_error( $results ) ) {
```
Comment (mukeshpanchal27) [inline src/wp-admin/includes/ajax-actions.php]: Both are same if not error then it's `true`. Why this change is needed?

Comment (costdev) [reply src/wp-admin/includes/ajax-actions.php]: Looks like a change for consistency. Technically-speaking, I'd tend to favour "If true then success", as anything other than true should not be a success. While it sounds the same, if `retrieve_password()` was ever broken and returned `false`, a success message would be sent.

**Review by audrasjb — APPROVED** [review]
This looks good to me.
@costdev pinging you for second thought: are you also OK with the logic proposed in this PR.

**Comment by costdev** [comment]
I think `true ===` is likely the safer option. With the PR's current logic, if `retrieve_password()` ever returned a non-true, non-WP_Error, the successful response would be sent. 

While `retrieve_password()` is documented for `true|WP_Error`, bugs are bugs, so `true ===` would at least flag that there's an issue since the successful response wouldn't be sent.

**Comment by LeTraceurSnork** [comment]
> I think `true ===` is likely the safer option. With the PR's current logic, if `retrieve_password()` ever returned a non-true, non-WP_Error, the email would be sent.
> 
> While `retrieve_password()` is documented for `true|WP_Error`, bugs are bugs, so `true ===` would at least flag that there's an issue since the email wouldn't be sent.

Have to disagree with you. Like my teamlead said once - "What, are we don't trust ourselves now?". We should trust our own code and if it says WP_Error or true - we considering not-WP_Error response as "true or identical". Ofc, if we want to extend function return to, say, WP_Error|bool (which is, btw, kinda weird - why is answers is so inconsistent?) - then we should also change its usages and return handlers - in the same PR (which is, in my opinion, should ot be this one, since that was not the original issue)

**Comment by costdev** [comment]
> "What, are we don't trust ourselves now?". We should trust our own code and if it says WP_Error or true - we considering not-WP_Error response as "true or identical"

This isn't so much about trusting our code as it is about being specific.

We want to check whether we can send a successful response:
- `if success then send a successful response` is specific.
- `if not a specific indicator of failure then send a successful response` is inverted logic that opens it up to other types of failure being missed.

If a successful response was ever sent when it shouldn't have been, we don't need to check the value of `$results` with `true === $results`. We know it's `true`. With `! is_wp_error( $results )`, we now need to investigate what's being returned. If anything other than `WP_Error` sends a success, we may not know there's a bug causing a non-`true|WP_Error` value to be returned for quite some time. Changing this to `true ===` means _as soon as_ there's a regression, something has behaved differently.

On the other topic of trusting our own code, your team lead and I would disagree on this. A discussion about those differing philosophies doesn't really belong on this PR, but if you want to discuss it further with me, feel free to DM me on Slack (same username) 🙂

**Review by alaminfirdows — APPROVED** [review]
Looks good to me.

File: src/wp-admin/includes/ajax-actions.php [hunk 3]
```diff
@@ -5571,7 +5571,12 @@ function wp_ajax_send_password_reset() {
...
 			sprintf( __( 'A password reset link was emailed to %s.' ), $user->display_name )
 		);
+	} elseif ( is_wp_error( $results ) ) {
+		wp_send_json_success(
+			/* translators: %s: User's display name. */
+			sprintf( __( 'A password reset link was emailed to %s.' ), $user->display_name )
+		);
 	} else {
-		wp_send_json_error( $results->get_error_message() );
+		wp_send_json_error( 'Cannot retrieve user\'s password', 500);
```
Comment (costdev) [inline src/wp-admin/includes/ajax-actions.php]: With the previous code here:
- If `retrieve_password()` returns `true`, a success is sent.
- If `retrieve_password()` returns something else, it's treated as the other documented return type: `WP_Error`, and an error is sent.
- In addition, if `retrieve_password()` is ever bugged and returns a non-`WP_Error`, the code will error out on `$results->get_error_message()`, highlighting the issue immediately. No "success" message will be sent _and_ the error will immediately alert the developer so they can rectify it. This way, documented return types are covered in a way that doesn't let anything else slip by unnoticed.

If we were to change the return type of `retrieve_password()`, then this change may be needed, but the return type is still `true|WP_Error` and the previous code here sufficiently covers these without allowing for any other value to get through (save for an object that has a `get_error_message()` method 😝).

File: src/wp-admin/users.php [hunk 4]
```diff
@@ -250,7 +250,7 @@
 
 			// Send the password reset link.
 			$user = get_userdata( $id );
-			if ( retrieve_password( $user->user_login ) ) {
+			if ( true === retrieve_password( $user->user_login ) ) {
```
Comment (costdev) [inline src/wp-admin/users.php:253]: I think this is the only change that's needed in this PR. It directly addresses the issue the ticket highlighted and I think we can discuss and reach consensus on consistency with these checks in a separate ticket.

Comment (costdev) [reply src/wp-admin/users.php:253]: Pinging @audrasjb and @SergeyBiryukov for their thoughts on whether we should keep this PR specific to resolving the bug, and separate out code quality improvements as they likely need some discussion to reach consensus.

Comment (LeTraceurSnork) [reply src/wp-admin/users.php:253]: @petitphp @costdev I've got two opposite comments. So, should I or should I not edit other lines than users.php:253?

p.s. I prefer to do that, so code quality would improve a bit

Comment (hellofromtonya) [reply src/wp-admin/users.php:253]: Hello 👋 

I agree with @costdev that only this fix is needed. I explained it in [more details in the Trac ticket](https://core.trac.wordpress.org/ticket/58407#comment:35).

Thank you for your contributions!

Comment (hellofromtonya) [reply src/wp-admin/users.php:253]: I'll close this PR when the change is committed. So no additional changes are needed in this PR.

**Comment by hellofromtonya** [comment]
The `true === retrieve_password()` approach to fix the reported bug was committed https://core.trac.wordpress.org/changeset/56937.
